### PR TITLE
Optimize UnsignedLong range queries to convert to MatchNoDocsQuery when lower > upper bounds

### LINKED ISSUE #14404: [Feature Request] Optimize `UnsignedLong` range queries to convert to `MatchNoDocsQuery` when lower > upper bounds
### Is your feature request related to a problem? Please describe

Coming from https://github.com/apache/lucene/pull/13425, we see that most number fields can take advantage of the optimization when lower value > upper value when doing a range query.

### Describe the solution you'd like

In https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/mapper/NumberFieldMapper.java#L1309, we should check if lower > upper and return a `MatchNoDocsQuery` if so.

### Related component

Search:Performance

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Skyring100**
### Description
unsignedLongRangeQuery method now returns MatchNoDocsQuery when lower bound > upper bound

### Related Issues
Resolves #14404 

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by Skyring100** [comment]
Thank your kind words and speedy response! This is my first time contributing to an open source project and I am a bit lost on how to update the test to make MatchNoDocsQuery a valid return type. 

I have looked inside the NumberFieldTypeTests class, am I supposed to change the lambda in the lambda in the second parameter to a different type? Or is there something about gradle I am supposed to change?

**Comment by harshavamsi** [comment]
> Thank your kind words and speedy response! This is my first time contributing to an open source project and I am a bit lost on how to update the test to make MatchNoDocsQuery a valid return type.
> 
> I have looked inside the NumberFieldTypeTests class, am I supposed to change the lambda in the second parameter to a different type? Or is there something about gradle I am supposed to change?

So digging into the test, we that in `NumberFieldTypeTests` the `testDocValueUnsignedLongRange` test is failing. `testDocValueUnsignedLongRange` internally calls [`doTestDocValueRangeQueries`](https://github.com/opensearch-project/OpenSearch/blob/178a7a073e1236d552f697e7f67d32f04a7fe8ec/server/src/test/java/org/opensearch/index/mapper/NumberFieldTypeTests.java#L654). We now see the error because on line 678, we only assert that it could be an instance of `IndexOrDocValuesQuery` but because we check against random values, there are cases where lower > upper and that would now return a `MatchNoDocsQuery`. So we would need to change https://github.com/opensearch-project/OpenSearch/blob/178a7a073e1236d552f697e7f67d32f04a7fe8ec/server/src/test/java/org/opensearch/index/mapper/NumberFieldTypeTests.java#L676C1-L678C129. First we should assert that it can be either `IndexOrDocValuesQuery` or `MatchNoDocsQuery`. 

So something like -- 
```java
assertThat(query, Matchers.either(instanceOf(IndexOrDocValuesQuery.class)).or(instanceOf(MatchNoDocsQuery.class)));
```

And then we should only check the `indexQuery` and `randomAccessQuery` parts if it happens to be an `IndexOrDocValuesQuery`

**Comment by harshavamsi** [comment]
@Skyring100 can you run `./gradlew spotlessApply` to run spotless and then commit changes?

**Review by harshavamsi — APPROVED** [review]
LGTM! @msfroh @reta take a look?

**Comment by Skyring100** [comment]
Thank you so much for your help @harshavamsi

File: server/src/test/java/org/opensearch/index/mapper/NumberFieldTypeTests.java [hunk 1]
```diff
@@ -73,6 +73,7 @@
 import org.opensearch.index.mapper.NumberFieldMapper.NumberType;
 import org.opensearch.index.query.QueryShardContext;
 import org.opensearch.search.MultiValueMode;
+import org.hamcrest.Matchers;
```
Comment (gaobinlong) [inline server/src/test/java/org/opensearch/index/mapper/NumberFieldTypeTests.java:76]: We can import `org.hamcrest.Matchers.either` only, better than the whole class.

`import static org.hamcrest.Matchers.either;`

**Comment by reta** [comment]
LGTM, thanks @Skyring100 ! Could you please resolved the conflicts? Thank you!

**Comment by Skyring100** [comment]
> LGTM, thanks @Skyring100 ! Could you please resolved the conflicts? Thank you!
I am not sure how to resolve these conflicts. I could not find the ResourceAwareTasksTests class mentioned in the stacktrace on my local machine. I have locally run ./gradlew precommit and ./gradlew spotlessApply, both of which were successful. Is there something about flaky tests affecting this issue?

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
@gaobinlong LGTM?

**Review by gaobinlong — APPROVED** [review]
Looks good to me.

**Comment by Skyring100** [comment]
Made the backport pull request though the command "git cherry-pick -x --mainline 1 d2c08b320b49920e142ec70fa66621305fd4ccf8" returned "fatal: bad object d2c08b320b49920e142ec70fa66621305fd4ccf8". Hopefully this doesn't cause an issue but provides some more information if there are errors for the request

**Comment by Skyring100** [comment]
I'd like to apologize for how many issues there have been, I was unaware of the knowledge needed beyond my programming skills. Thank you all for having patience and helping with the pull request.

**Pull request merged**